### PR TITLE
Fix (de)quantisation scale factor formula in MKL kernels

### PR DESCRIPTION
**Pull request by milpuz01**
This patch that fixes formula for calculating scale factor when MKL quantization and dequantization primitives are called from TF. This updated formulas match results from Eigen library and makes tests //tensorflow/python:quantized_ops_test and //tensorflow/python:dequantize_op_test pass when TF_ENABLE_ONEDNN_OPTS is 1 .

Additionally:
* it also adds correct memory layout when using tensor with different dimension to 4 in MKL dequantize operation, and
* reverts to Eigen op when per slice dequantization is specified

### CODE REVIEW

File: tensorflow/core/kernels/mkl/mkl_dequantize_op.cc [hunk 1]
```diff
@@ -53,15 +55,16 @@ class MklDequantizeOp : public OpKernel {
...
 
       // Get the inputs
-      const Tensor& src_tensor = MklGetInput(ctx, kSrcIndex);
-      const float min_range =
-          MklGetInput(ctx, kMinIndex).template flat<float>()(0);
-      const float max_range =
-          MklGetInput(ctx, kMaxIndex).template flat<float>()(0);
+      const Tensor& src_tensor = ctx->input(kSrcIndex);
+      const float min_range = ctx->input(1).flat<float>()(0);
+      const float max_range = ctx->input(2).flat<float>()(0);
```
Comment (mahmoud-abuzaina) [inline tensorflow/core/kernels/mkl/mkl_dequantize_op.cc:60]: Can you still use kMinIndex/kMaxIndex?

File: tensorflow/core/kernels/mkl/mkl_dequantize_op.cc [hunk 2]
```diff
@@ -53,15 +55,16 @@ class MklDequantizeOp : public OpKernel {
...
 
       // Get MklShape
       MklDnnShape src_mkl_shape;
       GetMklShape(ctx, kSrcIndex, &src_mkl_shape, native_format);
+      auto src_tf_shape = src_mkl_shape.IsMklTensor()
+                              ? src_mkl_shape.GetTfShape()
+                              : src_tensor.shape();
```
Comment (mahmoud-abuzaina) [inline tensorflow/core/kernels/mkl/mkl_dequantize_op.cc:67]: You can assume input is always NonMklTensor (just use shape() function). Other existing usages of IsMklTensor() will be cleaned up later.

Comment (mahmoud-abuzaina) [inline tensorflow/core/kernels/mkl/mkl_dequantize_op.cc:67]: You can assume input is always NonMklTensor (just use shape() function). Other existing usages of IsMklTensor() will be cleaned up later.

File: tensorflow/core/kernels/mkl/mkl_quantize_op.cc [hunk 3]
```diff
@@ -457,7 +457,7 @@ class MklQuantizeV2Op : public OpKernel {
         // If it is signed, we try to keep 0.0 being 0 and drop one bucket. For
         // example, if it is 8 bits, we have the range [-127, 127]. So for input
         // range of [-x, x], the scale should be 254/(2*x).
-        target_range = static_cast<float>((uint64_t{1} << (num_bits - 1)) - 1);
+        target_range = static_cast<float>((uint64_t{1} << num_bits) - 1) / 2.;
```
Comment (mahmoud-abuzaina) [inline tensorflow/core/kernels/mkl/mkl_quantize_op.cc:460]: Why is this change needed? tests seem to pass without this change. Also, this change caused some results mismatch in //tensorflow/core/kernels/mkl:mkl_quantize_op_test (this test only runs with --config=mkl).

Comment (mahmoud-abuzaina) [inline tensorflow/core/kernels/mkl/mkl_quantize_op.cc:460]: Why is this change needed? tests seem to pass without this change. Also, this change caused some results mismatch in //tensorflow/core/kernels/mkl:mkl_quantize_op_test (this test only runs with --config=mkl).

**Review by mahmoud-abuzaina — CHANGES_REQUESTED** [review]
Thank you for the fixes. Just had minor comments.

File: tensorflow/core/kernels/mkl/mkl_dequantize_op.cc (hunk 1 above)
Comment (milpuz01) [reply tensorflow/core/kernels/mkl/mkl_dequantize_op.cc]: Thanks for the comment. The code has been updated now to use `kMinIndex` and `kMaxIndex`

File: tensorflow/core/kernels/mkl/mkl_dequantize_op.cc (hunk 2 above)
Comment (milpuz01) [reply tensorflow/core/kernels/mkl/mkl_dequantize_op.cc]: Thanks for the comment. The code has been updated to remove reference to `src_mkl_shape` and it assumes that all tensors are not Mkl tensors.

File: tensorflow/core/kernels/mkl/mkl_quantize_op.cc (hunk 3 above)
Comment (milpuz01) [reply tensorflow/core/kernels/mkl/mkl_quantize_op.cc:460]: Without this change the test //tensorflow/python:quantized_ops_test fails. The following code is part of the test: 
```
#!/usr/bin/python3

import numpy as np
from tensorflow.python.framework import constant_op
from tensorflow.python.ops import array_ops
from tensorflow.python.framework import dtypes

values = np.array([-1, -0.5, 0.555, 1], dtype=np.float32)
quant_values = np.array([-128, -64, 71, 127], dtype=np.int32)

inputs = constant_op.constant(values)
expected_quantized = quant_values

min_range, max_range = -1.0, 0.8

quantized =  array_ops.quantize(
    inputs,
    min_range,
    max_range,
    T=dtypes.qint8,
    mode="SCALED",
    round_mode="HALF_TO_EVEN",
    axis=None).output

print('Returned: ', quantized.numpy())
print('Expected: ', expected_quantized)
```
Without above change the returned values are `[-127, -64, 70, 127]` instead of `[-128, -64, 71, 127]`. Also the above formula is formula that is here: https://intellabs.github.io/distiller/algo_quantization.html for calculating scale factor for full range.

Comment (milpuz01) [reply tensorflow/core/kernels/mkl/mkl_quantize_op.cc:460]: Please see above comment.

File: tensorflow/core/kernels/mkl/mkl_dequantize_op.cc [hunk 4]
```diff
@@ -78,49 +75,48 @@ class MklDequantizeOp : public OpKernel {
...
-        // There is no API in MKL-DNN v1.x to construct memory descriptor with
-        // same .data field but different type.
-        dst_md.data.data_type = memory::convert_to_c(MklDnnType<float>());
-      } else {
-        dst_md = memory::desc(src_dims, MklDnnType<float>(),
-                              src_dims.size() == 4 ? memory::format_tag::nhwc
-                                                   : memory::format_tag::nc);
-      }
+      memory::desc dst_md =
+          memory::desc(src_dims, MklDnnType<float>(), dst_layout_type);
```
Comment (mdfaijul) [inline tensorflow/core/kernels/mkl/mkl_dequantize_op.cc:114]: Could you please use `MklDnnType<T>()` instead of `MklDnnType<float>()`

**Comment by mahmoud-abuzaina** [comment]
I don't know why there are 2 copies of every comment I make. Sorry for that.

**Comment by ashiqimranintel** [comment]
Any update on this?

**Comment by milpuz01** [comment]
> Any update on this?

Hi @ashiqimranintel, I believe that all comments from reviewers were addresses. If there are any outstanding questions do let me know?

Thanks,
Milos.

**Comment by mahmoud-abuzaina** [comment]
@milpuz01 There was a comment I made few days ago but didn't show up for some reason. Pasting again here.
Thank for the explanation and the other changes. Can you then add this change to your PR to make the //tensorflow/core/kernels/mkl:mkl_quantize_op_test pass?

index ce35f5a979c..76155fad8df 100644
--- a/tensorflow/core/kernels/mkl/mkl_quantize_op_test.cc
+++ b/tensorflow/core/kernels/mkl/mkl_quantize_op_test.cc
@@ -74,7 +74,7 @@ TEST_F(MklQuantizeV2OpTest, small_int8) {
   Tensor expected(allocator(), DT_QINT8, TensorShape({8}));
   Tensor expected_min(allocator(), DT_FLOAT, TensorShape({}));
   Tensor expected_max(allocator(), DT_FLOAT, TensorShape({}));
-  test::FillValues<qint8>(&expected, {0, -1, 1, -2, -24, -128, -80, 127});
+  test::FillValues<qint8>(&expected, {0, -1, 1, -2, -25, -128, -81, 127});
   test::ExpectTensorEqual<qint8>(expected, *GetOutput(0));
   test::FillValues<float>(&expected_min, {-127.0});
   test::FillValues<float>(&expected_max, {127.0});```

**Comment by milpuz01** [comment]
> @milpuz01 There was a comment I made few days ago but didn't show up for some reason. Pasting again here. Thank for the explanation and the other changes. Can you then add this change to your PR to make the //tensorflow/core/kernels/mkl:mkl_quantize_op_test pass?

Thanks @mahmoud-abuzaina. Commit https://github.com/tensorflow/tensorflow/pull/56086/commits/5b680c934ecbd4fdb47158e3b91ab06be1b06e86 has the required change

**Comment by mahmoud-abuzaina** [comment]
@milpuz01 Thank you for the change. PR looks good to me now. Approved.

File: tensorflow/core/kernels/mkl/mkl_dequantize_op.cc [hunk 5]
```diff
@@ -78,49 +75,48 @@ class MklDequantizeOp : public OpKernel {
...
+          break;
+        case 4:
+          dst_layout_type = memory::format_tag::nhwc;
+          break;
+        case 5:
+          dst_layout_type = memory::format_tag::ndhwc;
+          break;
+        default:
+          OP_REQUIRES_OK(ctx,
+                         errors::Aborted("Input dims must be <= 5 and >= 1"));
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_dequantize_op.cc:97]: Nit: I think we can use invalid argument.
```suggestion
                         errors::InvalidArgument("Input dims must be <= 5 and >= 1"));
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you again for the PR and review! I'm so sorry for the delay! I have one additional minor comment.

**Comment by ashiqimranintel** [comment]
@milpuz01, can you address the above comment?

**Comment by milpuz01** [comment]
> @milpuz01, can you address the above comment?

Fixed!

**Review by penpornk — APPROVED** [review]
Thank you very much!

**Comment by okkwon** [comment]
Hi @milpuz01, this breaks the tensorflow integration with the IREE projects. Would you mind reverting/fixing/resubmitting it?

ERROR: /home/kbuilder/.cache/bazel/_bazel_kbuilder/1900d0fac5123d725c8d2e08b3f8c209/external/org_tensorflow/tensorflow/core/kernels/mkl/BUILD:280:22: Compiling tensorflow/core/kernels/mkl/mkl_dequantize_op.cc failed: (Exit 1): clang failed: error executing command /usr/lib/llvm-9/bin/clang -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 355 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox
external/org_tensorflow/tensorflow/core/kernels/mkl/mkl_dequantize_op.cc:59:53: error: use 'template' keyword to treat 'flat' as a dependent template name
      const float min_range = ctx->input(kMinIndex).flat<float>()(0);
                                                    ^
                                                    template
external/org_tensorflow/tensorflow/core/kernels/mkl/mkl_dequantize_op.cc:60:53: error: use 'template' keyword to treat 'flat' as a dependent template name
      const float max_range = ctx->input(kMaxIndex).flat<float>()(0);
                                                    ^
                                                    template
2 errors generated.

**Comment by penpornk** [comment]
@okkwon Sorry for the trouble! I tried fixing forward through https://github.com/tensorflow/tensorflow/commit/831e2d89dd2d29d673ee2781938251db287737be. Please let us know if there are any more errors. Thank you!
